coding-201-advanced-parsers/metaparser.py

```diff
@@ -0,0 +1,197 @@
+'''
+Metaparser
+----------
+   
+   Metaparser pyats module provides a structure unification among different
+   parsing contexts. It provides a unified structure that will be followed
+   by different device connections (Ex: 'cli', 'yang', 'xml).
+
+   It addresses the common automation problems as:
+   -----------------------------------------------
+
+   * Script changes due to CLI command and/or output changes
+     (release to release, OS to OS,  Platform/Series to Platform/Series) 
+   * Parser reuse limitation due to naming convention 
+
+'''
+
+import re
+import pprint
+from metaparser import MetaParser
+# Schema is the user datastructure that every output from the inheriting class
+# should comply with.
+	# * Any : Marks a section of a schema that matches anything.
+	# * Optional : Marks an optional part of the schema.
+	# * OR : Defines a schema of OR relationship, eg, the input data must pass
+	# the validation of one of the requirements of this Schema.
+from metaparser.util.schemaengine import Schema, Any, Optional, Or
+
+
+class ParserSchema(MetaParser):
+    # Schema could be any valid python datastructure/callable.
+    Schema = {
+        'vrf':
+            {Any():
+                {'neighbor':
+                    {Any():
+                        {'address_family':
+                            {Any():
+                            	# Optional here means that key is not a mandatory
+                            	# key in the parsed output.
+                                {Optional('route_map_in'): str,
+                            	# OR here means that parsed key could be either
+                            	# of type(str) or type(list).
+                                 'route_map_out': Or(str, list),
+                                }
+                            },
+                        }
+                    },
+                }
+            },
+        }
+
+
+class Parser(ParserSchema):
+
+    STRING = '''
```

Comment:
maybe leave this as docstring header

Reply:
If we need to have it as a running example, I can't have it as a docstring.